databuilder/file_formats/arrow.py

```diff
@@ -0,0 +1,87 @@
+import datetime
+from itertools import islice
+
+import pyarrow
+
+from databuilder.file_formats.validation import ValidationError, validate_headers
+
+PYARROW_TYPE_MAP = {
+    bool: pyarrow.bool_,
+    # If we added some kind of range specification as part of ColumnSpec we could use
+    # that to select smaller numeric types where appropriate
+    int: pyarrow.int64,
+    float: pyarrow.float64,
+    str: pyarrow.string,
+    datetime.date: pyarrow.date32,
+}
+
+# Reading around bits of old blog posts suggests that we want batches in roughly the
+# single digit megabyte range. Of course, translating this into a number of rows depends
+# on the width of the rows, which isn't constant given we use variable width types like
+# `string`. Plus we're using compression so the on disk size will vary in any case.
+# Given 30 columns, each of an average of 32 bits wide, then 64,000 rows takes about
+# 7.7MB, which seems roughly in the right ballpark.
```

Comment:
I think the bounds of "right" are pretty wide here. If we set it silly-low (like 10s of rows) then I imagine we'd see performance issues when writing and reading the files because of the amount of overhead we're adding. If we set it very high (like millions of rows) then we'd negate the benefits of streaming and see the memory consumption of databuilder jobs getting quite large.

Between those two bounds I suspect it makes very little practical difference. It's not like we're trying to optimise for the absolute maximum throughput or the absolute minimum memory consumption.

It's annoying we're forced to commit to, and justify, a number. If we were just dumping a Pandas frame to disk we wouldn't even be thinking about it. But unfortunately we can't just copy the default behaviour there (I tried) because it's based around bytes-per-chunk and we don't have straightforward access to that information.

Reply:
I've tried to make the comment slightly clearer here.